File: ts/packages/knowPro/src/storage.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { Batch } from "./common.js";
import {
    SemanticRef,
    SemanticRefOrdinal,
    IMessage,
    MessageOrdinal,
    ICollection,
    IStorageProvider,
    IMessageCollection,
    ISemanticRefCollection,
    IReadonlyCollection,
} from "./interfaces.js";

export class Collection<T, TOrdinal extends number>
    implements ICollection<T, TOrdinal>
{
    protected items: T[];

    constructor(items?: T[] | undefined) {
        this.items = items ?? [];
    }

    public get isPersistent(): boolean {
        return false;
    }

    public get length(): number {
        return this.items.length;
    }

    public get(ordinal: TOrdinal): T {
        return this.items[ordinal];
    }

    public getSlice(start: TOrdinal, end: TOrdinal): T[] {
        return this.items.slice(start, end);
    }

    public getMultiple(ordinals: TOrdinal[]): T[] {
        const items = new Array<T>(ordinals.length);
        for (let i = 0; i < ordinals.length; ++i) {
            items[i] = this.get(ordinals[i]);
        }
        return items;
    }

    public getAll(): T[] {
        return this.items;
    }

    public append(...items: T[]): void {
        for (const item of items) {
            this.items.push(item);
        }
    }

    public [Symbol.iterator](): Iterator<T, any, any> {
        return this.items[Symbol.iterator]();
    }
}

export class SemanticRefCollection
    extends Collection<SemanticRef, SemanticRefOrdinal>
    implements ISemanticRefCollection
{
    constructor(semanticRefs?: SemanticRef[]) {
        super(semanticRefs);
    }
}

export class MessageCollection<TMessage extends IMessage = IMessage>
    extends Collection<TMessage, MessageOrdinal>
    implements IMessageCollection<TMessage>
{
    constructor(messages?: TMessage[]) {
        super(messages);
    }
}

export class MemoryStorageProvider implements IStorageProvider {
    constructor() {}

    public createMessageCollection<
        TMessage extends IMessage = IMessage,
    >(): IMessageCollection<TMessage> {
        return new MessageCollection<TMessage>();
    }

    public createSemanticRefCollection(): ISemanticRefCollection {
        return new SemanticRefCollection();
    }

    public close(): void {}
}

export function* getBatchesFromCollection<T = any>(
    collection: IReadonlyCollection<T>,
    startAtOrdinal: number,
    batchSize: number,
): IterableIterator<Batch<T>> {
    let startAt = startAtOrdinal;
    while (true) {
        let batch = collection.getSlice(startAt, startAt + batchSize);
        if (batch.length === 0) {
            break;
        }
        yield { startAt, value: batch };
        startAt += batchSize;
    }
}

File: python/ta/typeagent/knowpro/storage.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from dataclasses import dataclass
from typing import (
    Any,
    Iterable,
    Iterator,
    Callable,
)

from .collections import (
    MemoryCollection,
    MemoryMessageCollection,
    MemorySemanticRefCollection,
)
from .convindex import ConversationIndex
from .convthreads import ConversationThreads
from .importing import MessageTextIndexSettings, RelatedTermIndexSettings
from .interfaces import (
    IConversationThreads,
    IMessage,
    IMessageTextIndex,
    IPropertyToSemanticRefIndex,
    IStorageProvider,
    ITermToRelatedTermsIndex,
    ITermToSemanticRefIndex,
    ITimestampToTextRangeIndex,
    JsonSerializer,
    MessageOrdinal,
    SemanticRef,
    SemanticRefOrdinal,
)
from .messageindex import MessageTextIndex
from .propindex import PropertyIndex
from .reltermsindex import RelatedTermsIndex
from .timestampindex import TimestampToTextRangeIndex


class MemoryStorageProvider[TMessage: IMessage](IStorageProvider[TMessage]):
    """A storage provider that operates in memory."""

    # Declare indexes as not-None but uninitialized - initialize() must be called
    _conversation_index: ConversationIndex
    _property_index: PropertyIndex
    _timestamp_index: TimestampToTextRangeIndex
    _message_text_index: MessageTextIndex
    _related_terms_index: RelatedTermsIndex
    _conversation_threads: ConversationThreads

    def __init__(
        self,
        message_text_settings: MessageTextIndexSettings,
        related_terms_settings: RelatedTermIndexSettings,
    ):
        # Store settings privately - no defaults, caller must provide
        self._message_text_settings = message_text_settings
        self._related_terms_settings = related_terms_settings
        # Indexes will be created in initialize()

    @classmethod
    async def create(
        cls,
        message_text_settings: MessageTextIndexSettings,
        related_terms_settings: RelatedTermIndexSettings,
    ) -> "MemoryStorageProvider[TMessage]":
        """Create and initialize a MemoryStorageProvider with all indexes."""
        instance = cls(message_text_settings, related_terms_settings)
        # Initialize all indexes using the provided settings
        instance._conversation_index = ConversationIndex()
        instance._property_index = PropertyIndex()
        instance._timestamp_index = TimestampToTextRangeIndex()
        instance._message_text_index = MessageTextIndex(message_text_settings)
        instance._related_terms_index = RelatedTermsIndex(related_terms_settings)
        # Use the same embedding settings for conversation threads
        thread_settings = message_text_settings.embedding_index_settings
        instance._conversation_threads = ConversationThreads(thread_settings)
        return instance

    async def get_conversation_index(self) -> ITermToSemanticRefIndex:
        return self._conversation_index

    async def get_property_index(self) -> IPropertyToSemanticRefIndex:
        return self._property_index

    async def get_timestamp_index(self) -> ITimestampToTextRangeIndex:
        return self._timestamp_index

    async def get_message_text_index(self) -> IMessageTextIndex[TMessage]:
        return self._message_text_index

    async def get_related_terms_index(self) -> ITermToRelatedTermsIndex:
        return self._related_terms_index

    async def get_conversation_threads(self) -> IConversationThreads:
        return self._conversation_threads

    async def create_message_collection(
        self,
        serializer: JsonSerializer[TMessage] | type[TMessage] | None = None,
    ) -> MemoryMessageCollection[TMessage]:
        """Create a new message collection."""
        if isinstance(serializer, JsonSerializer):
            raise ValueError("MemoryStorageProvider does not use a serializer.")
        return MemoryMessageCollection[TMessage]()

    async def create_semantic_ref_collection(self) -> MemorySemanticRefCollection:
        """Create a new semantic reference collection."""
        return MemorySemanticRefCollection()

    async def close(self) -> None:
        """Close the storage provider."""
        pass


# TODO: The rest of this file is not currently used.


@dataclass
class Batch[T]:
    """A batch of items from a collection."""

    start_at: int
    value: list[T]


async def get_batches_from_collection[T](
    collection: MemoryCollection[T, int],
    start_at_ordinal: int,
    batch_size: int,
) -> list[Batch[T]]:
    """Generate batches of items from a collection."""
    start_at = start_at_ordinal
    result = []
    while True:
        batch = await collection.get_slice(start_at, start_at + batch_size)
        if not batch:
            break
        result.append(Batch(start_at=start_at, value=batch))
        start_at += batch_size
    return result


async def map_collection[T](
    collection: MemoryCollection[T, int],
    callback: Callable[[T, int], T],
) -> list[T]:
    """Map a callback function over a collection."""
    results: list[T] = []
    size = await collection.size()
    for i in range(size):
        item = await collection.get_item(i)
        results.append(callback(item, i))
    return results


async def filter_collection[T](
    collection: MemoryCollection[T, int],
    predicate: Callable[[T, int], bool],
) -> list[T]:
    """Filter items in a collection based on a predicate."""
    results: list[T] = []
    size = await collection.size()
    for i in range(size):
        item = await collection.get_item(i)
        if predicate(item, i):
            results.append(item)
    return results
